SimTracker/TrackTriggerAssociation/plugins/TTTrackAssociator.cc

```diff
@@ -86,91 +81,89 @@ void TTTrackAssociator<Ref_Phase2TrackerDigi_>::produce(edm::Event& iEvent, cons
-        /// then go to the next track
+        /// If this TP does not appear in all stubs (allowing one wrong stub)
+        /// then try next TP.
         if (nnotfound > 1)
-          //if (!allFound)
           continue;
 
         /// If we are here, it means that the TrackingParticle
-        /// generates hits in all stubs but one of the current track
+        /// generates hits in all stubs (allowing one incorrect one) of the current track
         /// so put it into the vector
-        tpInAllStubs.push_back(iterAuxMap->first);
+        tpInAllStubs.push_back(auxPair.first);
       }
 
-      /// Count how many TrackingParticles we do have
+      /// Count how many TrackingParticles were associated to all stubs on this track.
+      /// FIX: Could avoid this by using std::set for tpInAllStubs?
       std::sort(tpInAllStubs.begin(), tpInAllStubs.end());
       tpInAllStubs.erase(std::unique(tpInAllStubs.begin(), tpInAllStubs.end()), tpInAllStubs.end());
       unsigned int nTPs = tpInAllStubs.size();
 
-      /// If only one TrackingParticle, GENUINE
-      /// if different than one, COMBINATORIC
+      /// If only one TP associated to all stubs (allowing one incorrect) on track: GENUINE or LOOSELY_GENUINE.
+      /// If 0 or >= 2 TP: COMBINATORIC
+      /// WARNING: This means if one TP matches all stubs, and another matches all STUBS except
+      /// one, then the trackToTrackingParticleMap will not be filled.
+      /// WARNING: This also means that trackToTrackingParticleMap will be filled if
+      /// one TP matches all stubs, except for an incorrect one in either PS or 2S modules.
       if (nTPs != 1)
         continue;
 
-      /// Here, the track may only be GENUINE
-      /// Fill the map
+      /// Here, the track may only be GENUINE/LOOSELY_GENUINE
+      /// CHECK: Surely if one incorrect PS stub, it can also be COMBINATORIC?
+      /// Fill the map associating track to its principle TP.
       trackToTrackingParticleMap.insert(std::make_pair(tempTrackPtr, auxMap.find(tpInAllStubs.at(0))->second));
 
     }  /// End of loop over Tracks
 
-    /// Clean the only map that needs cleaning
-    /// Prepare the output map wrt TrackingParticle
-    typename std::map<edm::Ptr<TrackingParticle>, std::vector<edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>>>>::iterator
-        iterMapToClean;
-    for (iterMapToClean = trackingParticleToTrackVectorMap.begin();
-         iterMapToClean != trackingParticleToTrackVectorMap.end();
-         ++iterMapToClean) {
+    /// Remove duplicates from the only output map that needs it.
+    /// (Map gets multiple entries per track if it has several stubs belonging to same TP).
+    for (auto& p : trackingParticleToTrackVectorMap) {
       /// Get the vector of edm::Ptr< TTTrack >
-      std::vector<edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>>> tempVector = iterMapToClean->second;
+      /// (CHECK: Couldn't this be done by reference, to save CPU?)
+      std::vector<TTTrackPtr> tempVector = p.second;
```

Comment:
avoid copy (as indicated in previous comments)

Reply:
done